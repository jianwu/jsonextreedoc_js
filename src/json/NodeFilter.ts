import { TDNode, TDNodeType } from "..";
import ListUtil from "../core/ListUtil";

/** if it returns undefined, node will be skipped */
abstract class NodeFilter {
  abstract apply(n: TDNode): TDNode | undefined;
  static exclude(...patterns: string[]): ExcludeFilter { return new ExcludeFilter(...patterns); }
  static mask(...patterns: string[]): MaskFilter { return new MaskFilter(...patterns); }
};

export default NodeFilter;

abstract class RegexFilter implements NodeFilter {
  public readonly pathPatterns:RegExp[] = [];
  abstract transform(n: TDNode): TDNode | undefined;
  public constructor(...strPatterns: string[]) { this.addPatterns(...strPatterns); }
  public apply(n: TDNode): TDNode | undefined {
    if (!this.matches(n.pathAsString))
      return n;
    return this.transform(n);
  }

  private addPatterns(...patterns: string[]) {
    if (patterns != null)
      for (const ptn of patterns)
        this.pathPatterns.push(new RegExp(ptn));
  }

  private matches(path: string): boolean {
    return ListUtil.exists(this.pathPatterns, p => p.test(path));
  }
}

class ExcludeFilter extends RegexFilter {
  public constructor(... patterns: string[]) { super(...patterns); }
  transform(n: TDNode): TDNode | undefined { return undefined; }
}

class MaskFilter extends RegexFilter {
  constructor(...patterns: string[]) {super(...patterns);}
  transform(n: TDNode): TDNode | undefined {
    if (n.value == null && !n.hasChildren())
      return n;
    return n.cloneOfSimpleType(this.getMaskStr(n));
  }
  private getMaskStr(n: TDNode): string  {
    switch (n.type) {
      case TDNodeType.SIMPLE: return "<Masked:len=" + n.value?.toString().length + ">";
      case TDNodeType.MAP: return "{Masked:size=" + n.getChildrenSize() + "}";
      case TDNodeType.ARRAY: return "[Masked:length=" + n.getChildrenSize() + "]";
    }
    return "[Masked]";  // Shouldn't happen
  }
}
