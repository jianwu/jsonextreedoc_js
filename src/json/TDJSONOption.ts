import { TDNodeType } from '../TDNode';

export class TDJSONOption {
  KEY_ID = `$id`;
  KEY_TYPE = "$type";
  deliminatorKey = ":";
  deliminatorValue = ",";
  deliminatorObjectStart = "{";
  deliminatorObjectEnd = "}";
  deliminatorArrayStart = "[";
  deliminatorArrayEnd = "]";
}
