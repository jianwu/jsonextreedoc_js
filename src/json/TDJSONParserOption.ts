import { TDNodeType } from '../TDNode';

export class TDJSONParserOption {
  public KEY_ID = `$id`;

  public uri?: string;

  /** In case there's no enclosed '[' of '{' on the root level, the default type. */
  public defaultRootType = TDNodeType.SIMPLE;

  public setDefaultRootType(type: TDNodeType) {
    this.defaultRootType = type;
    return this;
  }

 /**
  * if this is set, all the id in $id and $ref will be suffixed with "_" + docId, this is to avoid collision when merge
  * multiple docs as stream
  */
  docId?: string | number;

  public setDocId(id: string | number) {
    this.docId = id;
    return this;
  }
}
