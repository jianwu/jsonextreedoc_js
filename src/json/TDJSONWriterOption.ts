import { TDNode } from "..";
import { ValueType } from "../TDNode";
import NodeFilter from "./NodeFilter";

export default class TDJSONWriterOption {
  private mIndentFactor = 0;
  public alwaysQuoteName = true;
  public quoteChar = '"';
  /** @internal */
  public indentStr = ''; // Used internally
  
  /** Node mapper, if it returns null, node will be skipped */
  public nodeFilters: NodeFilter[] = [];

  public set indentFactor(indentFactor: number) {
    this.mIndentFactor = indentFactor;
    this.indentStr = ' '.repeat(this.mIndentFactor);
  }

  public get indentFactor() { return this.mIndentFactor}

  public setIndentFactor(indentFactor: number) {
    this.indentFactor = indentFactor;
    return this;
  }

  public setIndentStr(str: string) {
    this.indentStr = str;
    return this;
  }

  public setAlwaysQuoteName(alwaysQuoteName: boolean) {
    this.alwaysQuoteName = alwaysQuoteName;
    return this;
  }

  public setQuoteChar(quoteChar: string) {
    this.quoteChar = quoteChar;
    return this;
  }

  public hasIndent() {
    return this.indentStr.length > 0;
  }

  public applyFilters(n: TDNode): TDNode | undefined {
    let result: TDNode | undefined = n;
    for (const f of this.nodeFilters){
      if (result == null)
        break;
      result = f.apply(result);
    }
    return result;
  }

  public addNodeFilter(...filters: NodeFilter[]) {
    for (const f of filters)
      this.nodeFilters.push(f);
    return this;
  }
}
