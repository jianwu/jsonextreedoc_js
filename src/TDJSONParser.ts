import TDNode, { TDNodeType } from './TDNode';
import TDJSONParserOption from './TDJSONParserOption';
import CharSource from './CharSource';
import StringBuilder from './StringBuilder';

export default class TDJSONParser {
  public static readonly instance = new TDJSONParser();
  public static get() {
    return TDJSONParser.instance;
  }

  public parse(opt: TDJSONParserOption): TDNode {
    return this.parseFromSource(opt.source, opt, new TDNode());
  }

  public parseFromSource(src: CharSource, opt: TDJSONParserOption, node: TDNode): TDNode {
    if (!TDJSONParser.skipSpaceAndComments(src)) return node;

    const c = src.peek();
    node.start = src.getBookmark();
    try {
      if (c === '{') return this.parseMap(src, opt, node, true);

      if (c === '[') return this.parseArray(src, opt, node, true);

      if (node.isRoot()) {
        switch (opt.defaultRootType) {
          case TDNodeType.MAP:
            return this.parseMap(src, opt, node, false);
          case TDNodeType.ARRAY:
            return this.parseArray(src, opt, node, false);
        }
      }

      if (c === '"' || c === "'" || c === '`') {
        src.read();
        const sb = new StringBuilder();
        src.readQuotedToString(c, sb);
        this.readContinuousString(src, sb);
        return node.setValue(sb.toString());
      }

      const str = src.readUntilTermintor(',}]\n\r', 0, Number.MAX_VALUE).trim();
      if ('null' === str) return node.setValue(null);
      if ('true' === str) return node.setValue(true);
      if ('false' === str) return node.setValue(false);
      if (str.startsWith('0x') || str.startsWith('0X')) return node.setValue(this.parseNumber(str.substring(2), true));
      if (c === '-' || c === '+' || c === '.' || (c >= '0' && c <= '9'))
        return node.setValue(this.parseNumber(str, false));
      return node.setValue(str);
    } finally {
      node.end = src.getBookmark();
    }
  }

  private readContinuousString(src: CharSource, sb: StringBuilder): void {
    while (TDJSONParser.skipSpaceAndComments(src)) {
      const c = src.peek();
      if ('"`\''.indexOf(c) < 0) break;
      src.read();
      src.readQuotedToString(c, sb);
    }
  }

  /**
   * @return true if there's more text left
   */
  public static skipSpaceAndComments(src: CharSource): boolean {
    while (src.skipSpaces()) {
      const c = src.peek();
      if (c === '#') {
        if (src.skipUntilTerminator('\n')) src.skip(1);
        continue;
      }

      if (c !== '/' || src.isEof(1)) return true;
      const c1 = src.peek(1);
      switch (c1) {
        case '/': // line comments
          if (src.skipUntilTerminator('\n')) src.skip(1);
          break;
        case '*': // block comments
          src.skip(2);
          src.skipUntilMatch('*/', true);
          break;
        default:
          return true;
      }
    }
    return false;
  }

  public parseMap(src: CharSource, opt: TDJSONParserOption, node: TDNode, withStartBracket: boolean): TDNode {
    node.type = TDNodeType.MAP;
    if (withStartBracket) src.read();

    for (let i = 0; ; ) {
      if (!TDJSONParser.skipSpaceAndComments(src)) {
        if (withStartBracket) throw src.createParseRuntimeException("EOF encountered while expecting matching '}'");
        break;
      }

      let c = src.peek();
      if (c === '}') {
        src.read();
        break;
      }

      if (c === ',') {
        // Skip ,
        src.read();
        continue;
      }

      let key;
      if (c === '"' || c === "'" || c === '`') {
        src.read();
        key = src.readQuotedString(c);
        if (!TDJSONParser.skipSpaceAndComments(src)) break;
        c = src.peek();
        if (c !== ':' && c !== '{' && c !== '[' && c !== ',' && c !== '}')
          throw src.createParseRuntimeException("No ':' after key:" + key);
      } else {
        key = src.readUntilTermintor(':{[,}"', 1, Number.MAX_VALUE).trim();
        if (src.isEof()) throw src.createParseRuntimeException("No ':' after key:" + key);
        c = src.peek();
      }

      if (c === ',' || c === '}')
        // If there's no ':', we consider it as indexed value (array)
        node.createChild(i + '').setValue(key);
      else {
        if (c === ':') src.read();
        this.parseFromSource(src, opt, node.createChild(key));
      }
      i++;
    }
    return node;
  }

  private parseArray(src: CharSource, opt: TDJSONParserOption, node: TDNode, withStartBracket: boolean): TDNode {
    node.type = TDNodeType.ARRAY;
    if (withStartBracket) src.read();
    while (true) {
      if (!TDJSONParser.skipSpaceAndComments(src)) {
        if (withStartBracket) throw src.createParseRuntimeException("EOF encountered while expecting matching ']'");
        break;
      }

      const c = src.peek();
      if (c === ']') {
        src.read();
        break;
      }

      if (c === ',') {
        src.read();
        continue;
      }

      this.parseFromSource(src, opt, node.createChild());
    }
    return node;
  }

  private parseNumber(str: string, isHex: boolean): number | string {
    const num = !isHex && str.indexOf('.') >= 0 ? parseFloat(str) : parseInt(str, isHex ? 16 : 10);
    return Number.isNaN(num) ? str : num;
  }
}
