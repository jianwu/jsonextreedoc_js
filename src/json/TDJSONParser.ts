import { TDNode, TDNodeType } from '../TDNode';
import { TDJSONParserOption } from './TDJSONParserOption';
import { CharSource } from '../core/CharSource';
import { StringCharSource } from '../core/StringCharSource';
import { StringBuilder } from '../core/StringBuilder';
import { TreeDoc } from '../TreeDoc';
import { ClassUtil } from '../core/ClassUtil';
import { RecursivePartial } from '../core/LangUtil';
import { LangUtil } from '..';

const EOF = '\uFFFF';

export class TDJSONParser {
  public static readonly instance = new TDJSONParser();
  public static get() {
    return TDJSONParser.instance;
  }

  /** Parse all the JSON objects in the input stream until EOF and store them inside an root node with array type */
  public parseAll(src: CharSource | string, 
      option:RecursivePartial<TDJSONParserOption> = new TDJSONParserOption(), 
      node = new TreeDoc('root', option.uri).root, isRoot = true): TDNode {
    const opt = option instanceof TDJSONParserOption ? option : LangUtil.mergeDeep(new TDJSONParserOption(), option);

    if (typeof src === 'string')
      src = new StringCharSource(src);

    const doc = TreeDoc.ofArray();
    let docId = 0;
    while(src.skipSpacesAndReturnsAndCommas())
      TDJSONParser.get().parse(src, opt.setDocId(docId++), doc.root.createChild());
    return doc.root;
  }

  public parse(src: CharSource | string, 
      option: RecursivePartial<TDJSONParserOption> = new TDJSONParserOption(), 
      node = new TreeDoc('root', option.uri).root, isRoot = true): TDNode {
    const opt = option instanceof TDJSONParserOption ? option : LangUtil.mergeDeep(new TDJSONParserOption(), option);

    if (typeof src === 'string')
      src = new StringCharSource(src);

    const c = TDJSONParser.skipSpaceAndComments(src);
    if (c === EOF) 
      return node;

    try {
      node.start = src.getBookmark();

      if (c === '{')  
        return this.parseMap(src, opt, node, true);

      if (c === '[')
        return this.parseArray(src, opt, node, true);

      if (node.isRoot()) {
        switch (opt.defaultRootType) {
          case TDNodeType.MAP:
            return this.parseMap(src, opt, node, false);
          case TDNodeType.ARRAY:
            return this.parseArray(src, opt, node, false);
        }
      }

      if (c === '"' || c === "'" || c === '`') {
        src.read();
        const sb = new StringBuilder();
        src.readQuotedToString(sb, c);
        this.readContinuousString(src, sb);
        return node.setValue(sb.toString());
      }

      let term = opt.termValue;
      if (node.parent != null)
        // parent.type can either by ARRAY or MAP.
        term = node.parent.type === TDNodeType.ARRAY ? opt.termValueInArray : opt.termValueInMap;

      const str = src.readUntilTerminator(term, opt.termValueStrs).trim();
      return node.setValue(ClassUtil.toSimpleObject(str));
    } finally {
      node.end = src.getBookmark();
    }
  }

  private readContinuousString(src: CharSource, sb: StringBuilder): void {
    let c;
    while ((c = TDJSONParser.skipSpaceAndComments(src)) !== EOF) {
      if ('"`\''.indexOf(c) < 0)
        break;
      src.read();
      src.readQuotedToString(sb, c);
    }
  }

  /**
   * @return char next char to read (peeked), if '\uFFFF' indicate it's EOF
   */
  public static skipSpaceAndComments(src: CharSource): string {
    while (src.skipSpacesAndReturns()) {
      const c = src.peek();
      if (c === '#') {
        if (src.skipUntilTerminator('\n'))
          src.skip(1);
        continue;
      }

      if (c !== '/' || src.isEof(1)) return c;
      const c1 = src.peek(1);
      switch (c1) {
        case '/': // line comments
          if (src.skipUntilTerminator('\n'))
            src.skip(1);
          break;
        case '*': // block comments
          src.skip(2);
          src.skipUntilMatch('*/', true);
          break;
        default:
          return c1;
      }
    }
    return EOF;
  }

  public parseMap(src: CharSource, opt: TDJSONParserOption, node: TDNode, withStartBracket: boolean): TDNode {
    node.type = TDNodeType.MAP;
    if (withStartBracket)
      src.skip();

    for (let i = 0; ; ) {
      let c = TDJSONParser.skipSpaceAndComments(src);
      if (c === EOF) {
        if (withStartBracket)
          throw src.createParseRuntimeException("EOF while expecting matching '}' with '{' at " + node.start);
        break;
      }

      if (c === '}') {
        src.skip();
        break;
      }

      if (src.startsWith(opt.deliminatorValue)) { // Skip ,
        src.skip(opt.deliminatorValue.length);
        continue;
      }

      let key;
      if (c === '"' || c === "'" || c === '`') {
        src.skip();
        key = src.readQuotedString(c);
        c = TDJSONParser.skipSpaceAndComments(src);
        // if (c === EOF)
        //   break;
        if (!src.startsWith(opt.deliminatorKey) && c !== '{' && c !== '[' && c !== ',' && c !== '}')
          throw src.createParseRuntimeException(`No '${opt.deliminatorKey}' after key:${key}`);
      } else {
        key = src.readUntilTerminator(opt.termKey, opt.termKeyStrs, 1, Number.MAX_VALUE).trim();
        if (src.isEof())
          throw src.createParseRuntimeException(`No '${opt.deliminatorKey}' after key:${key}`);
        c = src.peek();
      }
      if (src.startsWith(opt.deliminatorKey))
        src.skip(opt.deliminatorKey.length);

      if (src.startsWith(opt.deliminatorValue) || c === '}')
        // If there's no ':', we consider it as indexed value (array)
        node.createChild(i + '').setValue(key);
      else {
        const childNode = this.parse(src, opt, node.createChild(key), false);
        if (opt.KEY_ID === key && childNode.type === TDNodeType.SIMPLE) {
          let id = childNode.value + '';
          if (opt.docId !== undefined) {
            id += "_" + opt.docId;
            childNode.value = id;
          }
          node.doc.idMap[id] = node;
        } else if (TDNode.REF_KEY === key && childNode.type === TDNodeType.SIMPLE) {
          if (opt.docId !== undefined)
            childNode.value += "_" + opt.docId;
        }   
      }
      i++;
    }
    return node;
  }

  private parseArray(src: CharSource, opt: TDJSONParserOption, node: TDNode, withStartBracket: boolean): TDNode {
    node.type = TDNodeType.ARRAY;
    if (withStartBracket)
      src.skip();
    while (true) {
      let c = TDJSONParser.skipSpaceAndComments(src);
      if (c === EOF) {
        if (withStartBracket)
          throw src.createParseRuntimeException("EOF while expecting matching ']' with '[' at " + node.start);
        break;
      }

      if (c === ']') {
        src.skip();
        break;
      }

      this.parse(src, opt, node.createChild(), false);
      c = TDJSONParser.skipSpaceAndComments(src);
      if (src.startsWith(opt.deliminatorValue)) {
        src.skip(opt.deliminatorValue.length);
      }    
    }
    return node;
  }
}
