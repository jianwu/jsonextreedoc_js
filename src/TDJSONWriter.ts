import TDJSONWriterOption from "./TDJSONWriterOption";
import TDNode, { TDNodeType } from "./TDNode";
import StringBuilder from "./StringBuilder";
import Appendable from "./Appendable";
import StringUtil from "./StringUtil";

export default class TDJSONWriter {
  public static readonly instance = new TDJSONWriter();
  public static get() { return TDJSONWriter.instance; }

  public writeAsString(node: TDNode, opt = new TDJSONWriterOption()): string {
    const out = new StringBuilder();
    this.write(out, node, opt);
    return out.toString();
  }

  public write(out: Appendable, node: TDNode, opt: TDJSONWriterOption, indentStr = ""): void {
    if (!node) {
      out.append("null");
      return;
    }

    const isCompact = opt.indentFactor === 0;
    let childIndentStr = "";
    if (!isCompact)
      childIndentStr = indentStr + opt.indentStr;

    switch (+node.type) {
      case TDNodeType.MAP:
        this.writeMap(out, node, opt, indentStr, childIndentStr);
        return;
      case TDNodeType.ARRAY:
        this.writeArray(out, node, opt, indentStr, childIndentStr);
        return;
      default:
        this.writeSimple(out, node, opt);
    }
  }

  private writeMap(out: Appendable, node: TDNode, opt: TDJSONWriterOption, indentStr: string, childIndentStr: string): void {
    out.append('{');
    if (node.children != null) {
      for (let i = 0; i < node.children.length; i++){
        const cn = node.children[i];
        if (opt.indentFactor > 0) {
          out.append('\n');
          out.append(childIndentStr);
        }
        if (!StringUtil.isJavaIdentifier(cn.key) || opt.alwaysQuoteName)  // Quote the key in case  it's not valid java identifier
          this.writeQuotedString(out, cn.key as string, opt.quoteChar);
        else
          out.append(cn.key as string);  // Map key will never be null
        out.append(":");
        this.write(out, cn, opt, childIndentStr);
        if (i < node.children.length - 1) // No need "," for last entry
          out.append(",");
      }

      if (opt.indentFactor > 0 && node.children.length > 0) {
        out.append('\n');
        out.append(indentStr);
      }
    }

    out.append('}');
  }

  private writeArray(out: Appendable, node: TDNode, opt: TDJSONWriterOption, indentStr: string, childIndentStr: string): void {
    out.append('[');
    if (node.children != null) {
      for (let i = 0; i < node.children.length; i++) {
        const cn = node.children[i];
        if (opt.indentFactor > 0) {
          out.append('\n');
          out.append(childIndentStr);
        }
        this.write(out, cn, opt, childIndentStr);
        if (i < node.children.length - 1) // No need "," for last entry
          out.append(",");
      }

      if (opt.indentFactor > 0 && node.children.length > 0) {
        out.append('\n');
        out.append(indentStr);
      }
    }

    out.append(']');
  }

  private writeSimple(out: Appendable, node: TDNode, opt: TDJSONWriterOption): void {
    if (typeof(node.value) === 'string') {
      this.writeQuotedString(out, node.value as string, opt.quoteChar);
      return;
    }

    out.append(node.value + "");
  }

  private writeQuotedString(out: Appendable, str: string, quoteChar: string): void {
    out.append(quoteChar);
    out.append(StringUtil.cEscape(str, quoteChar) as string);
    out.append(quoteChar);
  }
}
