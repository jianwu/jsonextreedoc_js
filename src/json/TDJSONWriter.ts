import { TDJSONWriterOption, TextType } from './TDJSONWriterOption';
import { TDNode, TDNodeType } from '../TDNode';
import { StringBuilder } from '../core/StringBuilder';
import { Appendable } from '../core/Appendable';
import { StringUtil } from '../core/StringUtil';
import { LangUtil } from '../core/LangUtil';

const NON_STRING = TextType.NON_STRING;
const STRING = TextType.STRING;
const KEY = TextType.KEY;
const OPERATOR = TextType.OPERATOR;

export class TDJSONWriter {
  public static readonly instance = new TDJSONWriter();
  public static get() {
    return TDJSONWriter.instance;
  }

  public static writeAsString(node: TDNode, opt: Partial<TDJSONWriterOption> = {}): string {
    return TDJSONWriter.get().writeAsString(node, opt);
  }

  public static write(out: Appendable, node: TDNode, opt: Partial<TDJSONWriterOption> = {}, indentStr = ''): Appendable {
    return TDJSONWriter.get().write(out, node, opt, (indentStr = ''));
  }

  public writeAsString(node: TDNode, opt: Partial<TDJSONWriterOption> = {}): string {
    const out = new StringBuilder();
    this.write(out, node, opt);
    return out.toString();
  }

  public write(out: Appendable, node: TDNode, option: Partial<TDJSONWriterOption> = {}, indentStr = ''): Appendable {
    const opt = LangUtil.mergeDeep(new TDJSONWriterOption(), option);
    if (!node)
      return  out.append(opt.deco("null", NON_STRING))

    let childIndentStr = '';
    if (opt.hasIndent())
      childIndentStr = indentStr + opt.indentStr;

    switch (node.type) {
      case TDNodeType.MAP:
        return this.writeMap(out, node, opt, indentStr, childIndentStr);
      case TDNodeType.ARRAY:
        return this.writeArray(out, node, opt, indentStr, childIndentStr);
      default:
        return this.writeSimple(out, node, opt);
    }
  }

  private writeMap(out: Appendable, node: Readonly<TDNode>, opt: TDJSONWriterOption, indentStr: string, childIndentStr: string): Appendable {
    out.append(opt.deco("{", OPERATOR));
    if (node.children != null) {
      for (let i = 0; i < node.getChildrenSize(); i++) {
        const cn = opt.applyFilters(node.children[i]);
        if (cn == null)
          continue;

        if (opt.hasIndent()) {
          out.append('\n');
          out.append(childIndentStr);
        }
        if (!StringUtil.isJavaIdentifier(cn.key) || opt.alwaysQuoteName)
          // Quote the key in case  it's not valid java identifier
          this.writeQuotedString(out, cn.key as string, opt, KEY);
        else 
          out.append(opt.deco(cn.key as string, KEY)); // Map key will never be null
        out.append(opt.deco(":", OPERATOR));
        this.write(out, cn, opt, childIndentStr);
        if (i < node.getChildrenSize() - 1)
          // No need "," for last entry
          out.append(opt.deco(",", OPERATOR));
      }

      if (opt.hasIndent() && node.hasChildren()) {
        out.append('\n');
        out.append(indentStr);
      }
    }

    return out.append(opt.deco("}", OPERATOR));
  }

  private writeArray(out: Appendable, node: Readonly<TDNode>, opt: TDJSONWriterOption, indentStr: string, childIndentStr: string): Appendable {
    out.append(opt.deco("[", OPERATOR));
    if (node.children != null) {
      for (let i = 0; i < node.getChildrenSize(); i++) {
        const cn = node.children[i];
        if (opt.hasIndent()) {
          out.append('\n');
          out.append(childIndentStr);
        }
        this.write(out, cn, opt, childIndentStr);
        if (i < node.getChildrenSize() - 1)
          // No need "," for last entry
          out.append(opt.deco(",", OPERATOR));
      }

      if (opt.hasIndent() && node.children.length > 0) {
        out.append('\n');
        out.append(indentStr);
      }
    }

    return out.append(opt.deco("]", OPERATOR));
  }

  private writeSimple(out: Appendable, node: TDNode, opt: TDJSONWriterOption): Appendable {
    const value = node.value;
    if (typeof value === 'string')
      return this.writeQuotedString(out, value as string, opt, STRING);

    return out.append(opt.deco(value + '', NON_STRING));
  }

  private writeQuotedString(out: Appendable, str: string, opt: TDJSONWriterOption, textType: TextType): Appendable {
    out.append(opt.quoteChar);
    out.append(opt.deco(StringUtil.cEscape(str, opt.quoteChar) as string, textType));
    return out.append(opt.quoteChar);
  }
}
