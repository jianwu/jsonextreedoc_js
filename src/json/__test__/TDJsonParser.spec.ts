import { StringCharSource } from '../../core/StringCharSource';
import { TDJSONParser } from '../../json/TDJSONParser';
import { TDJSONParserOption } from '../../json/TDJSONParserOption';
import { TDNodeType } from '../../TDNode';
import { TDJSONWriter } from '../../json/TDJSONWriter';
import { TDJSONWriterOption } from '../../json/TDJSONWriterOption';
import { JSONPointer } from '../../json/JSONPointer';
import { TDNode } from '../..';
import { TreeDoc_merge } from '../../TreeDoc';
import { TestData } from './TestData';

const testData = new TestData()

describe('TDJsonParser', () => {
  test('testSkipSpaceAndComments', () => {
    let src = new StringCharSource('  //abcd \n // defghi \n abc');
    expect(TDJSONParser.skipSpaceAndComments(src)).toBe('a');
    expect(src.readString(3)).toBe('abc');

    src = new StringCharSource('  #abcd \n # defghi \n abc');
    expect(TDJSONParser.skipSpaceAndComments(src)).toBe('a');
    expect(src.readString(3)).toBe('abc');

    src = new StringCharSource('/* abcd*/ \n /* defghi*/ \n abc');
    expect(TDJSONParser.skipSpaceAndComments(src)).toBe('a');
    expect(src.readString(3)).toBe('abc');
  });

  test('testParse', () => {
    const node = TDJSONParser.get().parse(testData.testData);
    const json = TDJSONWriter.get().writeAsString(node);

    console.log(`testParse:json=${json}`);

    expect(node.getChildValue('3')).toBe('valueWithoutKey');
    expect(node.getChildValue('6')).toBe('lastValueWithoutKey');
    expect(node.getChildValue('limit')).toBe(10);
    expect(node.getChildValue('total')).toBe('100000000000000000000');
    expect(node.getChildValue('maxSafeInt')).toBe(9007199254740991);
    expect(node.getValueByPath('data/0/name')).toBe('Some Name 1');
    expect(node.getValueByPath('data/1/address/streetLine')).toBe('2nd st');
    const n = node.getByPath('data/1');
    expect(node.getByPath('data/1')!.key).toBe('1');
    expect(node.getByPath(['data', '1'])!.key).toBe('1');

    expect(node.getChild('total')!.isLeaf()).toBeTruthy();
    expect(node.getChild('data')!.isLeaf()).toBeFalsy();
    expect(node.getByPath('data/1')!.path).toEqual(['data', '1']);

    const node1 = TDJSONParser.get().parse(json);
    expect(TDJSONWriter.get().writeAsString(node1)).toBe(json);
  });

  test('testParseProto', () => {
    const node = TDJSONParser.get().parse(testData.proto, new TDJSONParserOption().setDefaultRootType(TDNodeType.MAP));
    const json = TDJSONWriter.get().writeAsString(node,
      new TDJSONWriterOption().setIndentFactor(2).setAlwaysQuoteName(false),
    );
    console.log(`testParseProto:json=${json}`);
    expect(node.getValueByPath('n/n1/0/n11/1/n111')).toBeFalsy();
    expect(node.getValueByPath('n/n1/1/[d.e.f]')).toBe(4);
    expect(node.getValueByPath('n/n3/0')).toBe(6);
    expect(node.getByPath('n/n1/0')!.key).toBe('0');
    expect(node.getByPath('n/n1/1')!.key).toBe('1');
  });

  test('testParseJson5', () => {
    const node = TDJSONParser.get().parse(testData.JSON5);
    const json = TDJSONWriter.get().writeAsString(node,
      new TDJSONWriterOption().setIndentFactor(2).setAlwaysQuoteName(false),
    );
    console.log(`testParseJson5:json=${json}`);
    expect(node.getValueByPath('unquoted')).toBe('and you can quote me on that');
    expect(node.getValueByPath('hexadecimal')).toBe(912559);
    expect(node.getValueByPath('leadingDecimalPoint')).toBe(0.8675309);
    expect(node.getValueByPath('positiveSign')).toBe(1);
  });

  test('testRootMap', () => {
    const node = TDJSONParser.get().parse("'a':1\nb:2", new TDJSONParserOption().setDefaultRootType(TDNodeType.MAP));
    expect(node.getValueByPath('a')).toBe(1);
    expect(node.getValueByPath('b')).toBe(2);
  });

  test('testRootArray', () => {
    const testDataRootArray = `
    ,
    1
    2
    {
      v: 3
    },
    ,
    4,
    ,
    `;
    const node = TDJSONParser.get().parse(testDataRootArray, new TDJSONParserOption().setDefaultRootType(TDNodeType.ARRAY));
    const json = TDJSONWriter.get().writeAsString(node, new TDJSONWriterOption().setIndentFactor(2).setAlwaysQuoteName(false));
    console.log(`testParseJson5:json=${json}`);
    expect(node.getChildrenSize()).toBe(7);
    expect(node.getValueByPath('2')).toBe(2);
    expect(node.getValueByPath('3/v')).toBe(3);
  });

  test('testInvalid', () => {
    let node = TDJSONParser.get().parse('}');
    expect(node.value).toBe('}');

    node = TDJSONParser.get().parse('');
    expect(node.value).toBeNull();

    node = TDJSONParser.get().parse('[}]');
    expect(node.getChild(0)!.value).toBe('}');
  });

  test('testTDPath', () => {
    const jp = JSONPointer.get();
    const node = TDJSONParser.get().parse(testData.testData);
    const node1 = jp.query(node, '#1') as TDNode;
    expect(node1.getChildValue('name')).toBe('Some Name 1');
    // Relative with number support removed
    // expect(jp.query(node1, '2/limit')!.value).toBe(10);
    expect(node.value).toBeNull();
  });

  test('TDNode.toString', () => {
    const node = TDJSONParser.get().parse(testData.testData, new TDJSONParserOption().setDefaultRootType(TDNodeType.MAP));
    const str = node.toString();
    console.log('testToString:str=' + str);
    const str1 = node.toString();
    expect(str1).toBe(str);
    const city = node.getByPath('/data/0/address/city')! as TDNode;
    expect(city).not.toBeNull();
    city.setValue(city.value);
    const str2 = node.toString();
    // expect(str2).not.toBe(str);  // Javascript doesn't have real identity equality operator to verify it.
    expect(str2).toEqual(str);

    city.setValue('other city');
    node.freeze();
    const str3 = node.toString();
    console.log('testToString:str=' + str3);

    // toString should return different value when node value changed
    expect(str3).not.toEqual(str);

    expect(str3).toEqual("{total: '100000000000000000000', maxSafeInt: 9007199254740991, limit: 10, 3: 'valueWithoutKey', data: [{$id: '1', name: 'Some Name 1', address: {streetLine: '1st st', city: 'other city'}, createdAt: '2017-07-14T17:17:33.010Z', ip: '10.1.22.22'}, {$id: '2', name: 'Some Name 2', address: {streetLine: '2nd st', city: 'san jose'}, createdAt: '2017-07-14T17:17:33.010Z'}, 'Multiple line literal\\n    Line2'], objRef: {$ref: '1'}, 6: 'lastValueWithoutKey'}");

    const strWithoutRootKey = node.getChild("data")!.toStringInternal('', false, false, 100).toString();
    expect(strWithoutRootKey).toEqual("[{name: 'Some Name 1', address: {streetLine: '1st st', city: 'other city'}, createdAt: '2017-07-14T17...', ...}, ...]");

    const strWithoutRootKeyLimited = node.getChild("data")!.toStringInternal('', false, false, 10).toString();
    expect(strWithoutRootKeyLimited).toEqual("[{name: 'So...', ...}, ...]");
  });

  test('testStream', () => {
    const reader = new StringCharSource(testData.stream);
    const nodes: TDNode[] = [];
    while(reader.skipSpacesAndReturnsAndCommas())
      nodes.push(TDJSONParser.get().parse(reader));
    const node = TreeDoc_merge(nodes).root;
    console.log("testStream=" + node.toString());
    expect(node.getChild(1)?.key).toEqual("1");
    expect(node.getChild(1)?.getChild(0)?.doc).toBe(node.doc);
    expect(node.toString()).toMatchSnapshot();
  });

  test('testParseAll', () => {    
    let node = TDJSONParser.get().parseAll(testData.stream);
    console.log("testStream=" + node.toString());
    expect(node.toString()).toMatchSnapshot();

    const docFirstElement = node.doc.retain(node.children![0]);
    node = docFirstElement.root;
    console.log("testStream=" + node.toString());
    expect(node.key).toEqual("root");
    expect(node.toString()).toMatchSnapshot();
  });
});
