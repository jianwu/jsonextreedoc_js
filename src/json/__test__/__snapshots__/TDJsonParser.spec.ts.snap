// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TDJsonParser testParseAll 1`] = `"[{a: 1, obj: {$id: '1_0'}, ref: {$ref: '#1_0'}}, {b: 2, obj: {$id: '1_1'}, ref: {$ref: '#1_1'}}, 'a:1', 'b:2']"`;

exports[`TDJsonParser testParseAll 2`] = `"{a: 1, obj: {$id: '1_0'}, ref: {$ref: '#1_0'}}"`;

exports[`TDJsonParser testStream 1`] = `"[{a: 1, obj: {$id: '1_0'}, ref: {$ref: '#1_0'}}, {b: 2, obj: {$id: '1_1'}, ref: {$ref: '#1_1'}}, 'a:1', 'b:2']"`;
